File: app/src/main/java/br/com/mov/views/recyclerview/adapter/MovieAdapter.java
package br.com.mov.views.recyclerview.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.Filterable;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import br.com.mov.R;
import br.com.mov.models.Movie;

public class MovieAdapter extends RecyclerView.Adapter<MovieAdapter.MyViewHolder> implements Filterable {

    private Context context;
    private List<Movie> movies;
    private List<Movie> movieListFull;

    //regionFilter
    private Filter filter = new Filter() {
        @Override
        protected FilterResults performFiltering(CharSequence constraint) {
            List<Movie> filteredList = new ArrayList<>();

            if (constraint == null || constraint.length() == 0) {
                filteredList.addAll(movieListFull);
            } else {
                String filterPattern = constraint.toString().toLowerCase().trim();

                for (Movie movie : movieListFull) {
                    if (movie.getTitle().toLowerCase().contains(filterPattern)) {
                        filteredList.add(movie);
                    }
                }
            }

            FilterResults results = new FilterResults();
            results.values = filteredList;

            return results;
        }

        @Override
        protected void publishResults(CharSequence constraint, FilterResults results) {
            movies.clear();
            movies.addAll((List) results.values);
            notifyDataSetChanged();
        }
    };

    @Override
    public Filter getFilter() {
        return filter;
    }
    //endregion

    public MovieAdapter(Context context, List<Movie> movies) {
        this.context = context;
        this.movies = movies;
        movieListFull = new ArrayList<>(movies);
    }

    @NonNull
    @Override
    public MyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        View view = LayoutInflater.from(context).inflate(R.layout.item_movie, parent, false);
        return new MyViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull MyViewHolder holder, int position) {

        holder.movieTitle.setText(movies.get(position).getTitle());
        holder.movieImg.setImageResource(movies.get(position).getThumbnail());

    }

    @Override
    public int getItemCount() {
        return movies.size();
    }

    class MyViewHolder extends RecyclerView.ViewHolder {

        private TextView movieTitle;
        private ImageView movieImg;

        MyViewHolder(@NonNull View itemView) {
            super(itemView);

            movieTitle = itemView.findViewById(R.id.item_movie_title);
            movieImg = itemView.findViewById(R.id.item_movie_img);
        }
    }
}

File: app/src/main/java/br/com/mov/views/recyclerview/adapter/MovieAdapter.kt
package br.com.mov.views.recyclerview.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import br.com.mov.R
import br.com.mov.extensions.loadThumbnail
import br.com.mov.models.dto.MovieRequest
import java.util.*

class MovieAdapter(
        private val context: Context,
        private var movies: List<MovieRequest>,
        var onItemCliclListener:(movie: MovieRequest) -> Unit = {}
) : RecyclerView.Adapter<MovieAdapter.MyViewHolder>(), Filterable {

    private var movieListFull: List<MovieRequest> = movies.toList()

    //regionFilter
    private val filter = object : Filter() {
        override fun performFiltering(constraint: CharSequence?): FilterResults {
            val filteredList: MutableList<MovieRequest> = mutableListOf()

            if (constraint.isNullOrEmpty()) {
                filteredList.addAll(movieListFull)
            } else {
                val filterPattern = constraint.toString().toLowerCase(Locale.getDefault()).trim()

                for (movie in movieListFull) {
                    if (movie.title?.toLowerCase(Locale.getDefault())?.contains(filterPattern)!!) {
                        filteredList.add(movie)
                    }
                }
            }

            val results = FilterResults()
            results.values = filteredList
            results.count = filteredList.size

            return results
        }

        override fun publishResults(constraint: CharSequence, results: FilterResults) {
            movies = (results.values as List<*>).filterIsInstance<MovieRequest>() as MutableList<MovieRequest>
            notifyDataSetChanged()
        }
    }

    override fun getFilter(): Filter {
        return filter
    }
    //endregion

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): MyViewHolder {

        val view = LayoutInflater.from(context).inflate(R.layout.item_movie, parent, false)
        return MyViewHolder(view)
    }

    override fun onBindViewHolder(holder: MyViewHolder, position: Int) {

        holder.bind(movies[position])

    }

    override fun getItemCount() = movies.size

    inner class MyViewHolder(itemView: View) :
            RecyclerView.ViewHolder(itemView) {

        private lateinit var movie: MovieRequest
        val movieTitle: TextView = itemView.findViewById(R.id.item_movie_title)
        val movieImg: ImageView = itemView.findViewById(R.id.fragment_movie_detail_item_movie_img)

        fun bind(movie: MovieRequest) {
            this.movie = movie
            movieTitle.text = movie.title
            movie.pictureUrl?.let { movieImg.loadThumbnail(it) }
        }

        init {
            itemView.setOnClickListener {
                if (::movie.isInitialized) {
                    onItemCliclListener(movie)
                }
            }
        }

    }

}
